### PR TITLE
Electrochemistry! (finally) a.k.a. RMG-electrocat

### PR DESCRIPTION
**Pull request by JacksonBurns**
# Electrochemistry!

This PR finally brings electrochemistry to RMG.
It replaces or combines #2000 and  #2316.
Most of the original work was done by @davidfarinajr and @mjohnson541.
Help and effort rebasing from @JacksonBurns and @ssun30.
With some assistance along the way by @rwest.

Over the years it been checked many times. All the tests now pass.
It goes with RMG-database pull request https://github.com/ReactionMechanismGenerator/RMG-database/pull/667 
and it uses the [marcus_development](https://github.com/ReactionMechanismGenerator/ReactionMechanismSimulator.jl/tree/marcus_development) branch of RMS (which it checks out during the CI).
_The manual installation instructions need tweaking to help people install that branch, or we need to merge that branch to main in RMS._

### CODE REVIEW

**Comment by JacksonBurns** [comment]
@rwest @mjohnson541 the remaining unit test failures (here: https://github.com/ReactionMechanismGenerator/RMG-Py/actions/runs/7750195947/job/21136056656#step:10:2072) seem related to the echem work, and not anything in this PR. Unfortunately (see below) we may have to move `electrochem` development to this branch, or else `reset` `electrochem` to the latest commit on this branch so it looks like the rebase was done on `electrochem` directly?

Unfortunately it seems that I have yet again misunderstood how on earth git works. I think if I had _merged_ `main` into this branch, we could have then merged _this branch_ into the `electrochem` branch, and then the `electrochem` branch into `main`. The way I have done it here has basically stacked all the commits to `main` from before the last time `electrochem` was rebased, and then the commits on `electrochem`, and then my commits to make the tests work. I guess the history would look wacky if I did it _not_ this way?

**Comment by mjohnson541** [comment]
Thanks @JacksonBurns! This is great! I should be able to consolidate this with the last parts of echem development I have locally soon and then it should be all together for review.

**Comment by rwest** [comment]
I rebased this onto the latest main, and changed the base (target) for the PR back to main. Also rebased the `lithium` branch in the RMG-database which this refers to. Let's see how the tests do...

File: rmgpy/data/kinetics/library.py [hunk 1]
```diff
@@ -219,6 +219,30 @@ def get_sticking_coefficient(self, T):
             
         return False
 
+    def apply_solvent_correction(self, solvent):
```
Comment (rwest) [inline rmgpy/data/kinetics/library.py:222]: Has all of this solvent corrections to kinetics been done on a different branch (and reviewed independently?) or is it all blurred in with the electrochemistry stuff?

Comment (mjohnson541) [reply rmgpy/data/kinetics/library.py:222]: So I had a solvation correction system I built for the electrochemistry work. Yunsie was also working on solvation corrections and had her own dataset and a system that was similar, but also a bit different. So I modified my system to merge the two systems and handle both my dataset and her dataset.

Comment (rwest) [reply rmgpy/data/kinetics/library.py:222]: Did hers never make it into RMG then? this is the [only] chance to review both?

Comment (mjohnson541) [reply rmgpy/data/kinetics/library.py:222]: Yes, this is all only here.

**Comment by mjohnson541** [comment]
I just added the additional changes on top of this branch.

**Comment by mjohnson541** [comment]
I have also updated the database branch.

File: examples/rmg/CO2RR/input.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    surfPotential=(-1.2,'V'),
+    initialConcentrations={
+        # "CO2": (1e-3,'mol/cm^3'),
+        "proton": (1e-3,'mol/m^3'),
+        # "water": (0.055, 'mol/cm^3'),
+    },
+	initialSurfaceCoverages={
+        # "HX": 0.5,
+        # # "CXO2": 0.0,
+        "CHO2X": 1.0,
```
Comment (mjohnson541) [inline examples/rmg/CO2RR/input.py]: Are you unable to start from a vacant surface for some reason?

File: rmgpy/rmg/model.py [hunk 3]
```diff
@@ -811,7 +811,11 @@ def process_new_reactions(self, new_reactions, new_species, pdep_network=None, g
         Makes a reaction and decides where to put it: core, edge, or PDepNetwork.
         """
         for rxn in new_reactions:
-            rxn, is_new = self.make_new_reaction(rxn, generate_thermo=generate_thermo, generate_kinetics=generate_kinetics)
+            try:
+                rxn, is_new = self.make_new_reaction(rxn, generate_thermo=generate_thermo, generate_kinetics=generate_kinetics)
+            except Exception as e:
+                logging.error(f"Error when making reaction {rxn} from {rxn.family}")
+                raise
```
Comment (mjohnson541) [inline rmgpy/rmg/model.py:818]: Shouldn't this be 'raise e'?

File: test/database/databaseTest.py [hunk 4]
```diff
@@ -132,7 +132,9 @@ def test_kinetics(self):
                     ), "Kinetics surface family {0}: entries have surface attributes?".format(family_name)
                 if family_name not in {
                     "Surface_Proton_Electron_Reduction_Alpha",
+                    "Surface_Proton_Electron_Reduction_Alpha_vdW",
```
Comment (mjohnson541) [inline test/database/databaseTest.py:135]: Is there a reason for this? Can it be included in the commit message?

File: rmgpy/data/kinetics/family.py [hunk 5]
```diff
@@ -1519,10 +1519,9 @@ def apply_recipe(self, reactant_structures, forward=True, unique=True, relabel_a
             # If product structures are Group objects and the reaction is in certain families
             # (families with charged substances), the charge of structures will be updated
             if isinstance(struct, Molecule):
-                struct.update_charge()
```
Comment (mjohnson541) [inline rmgpy/data/kinetics/family.py]: Is it possible for us to instead skip updating charge for just the cutting labels and not the whole fragment?

**Review by mjohnson541 — COMMENTED** [review]
I went over all of the new commits. Mostly little things.

**Comment by rwest** [comment]
In order to get this in we should either tell everyone to use the marcus_development branch of RMS by tweaking the installation instructions like this
Screenshot 2024-11-06 at 10 58 35 AM

...or merge that branch to main on RMS first. Which I think @ssun30 is opening a PR for.

**Comment by ssun30** [comment]
Hi all, @rwest @JacksonBurns @mjohnson541 

After some discussion with Richard, we laid out the following list of actions for deploying this PR and [its sister PR on RMG-database](https://github.com/ReactionMechanismGenerator/RMG-database/pull/667):
- [ ] Test and make sure the `lithium_rebase` branch of RMG-database passes our CI workflow, with the `echem_rebase` branch of RMG-Py and `marcus_development` branch of RMS. 
- [ ] Test and make sure the `echem-rebase` branch of RMG-Py passes our CI workflow, with the `lithium_rebase` branch of RMG-database and `marcus_development branch of RMS. 
- [ ] Revert the CI workflow of `lithium_rebase` back to using `main` branch of RMG-Py. This **_will_** make the CI fail, so we should override the automerge.  
- [ ] Merge the `lithium_rebase` branch into `main` of RMG-database. 
- [ ] Revert the CI workflow of `echem-rebase` back to using `main` branch of RMG-database. This **_will not_** fail since the database branch was already merged. 
- [ ] Merge the `echem-rebase` branch into `main` of RMG-Py.
- [ ] After successfully merging both branches, I will open a new PR that reverts the RMS branch to `main`. But this will remain a WIP until we merge #2687 

The echem PR has been pretty mature and stable at this point, we hope we could have it deployed by the end of the week.

**Comment by JacksonBurns** [comment]
So after you finish this checklist will users be able to use RMG-electrocat? Or will they have to wait for the development version of RMS that you are using to be merged into RMS main?

I will note that the PR you are waiting on (#2687) is currently mostly in progress by me in #2694, and I not able to get the new `juliacall` code to work. I have currently made RMS optional (see #2735) and intend to just get the Python 3.9 upgrade to work first, and then go back and add RMS back after.

Perhaps we need to Zoom...

**Comment by rwest** [comment]
> So after you finish this checklist will users be able to use RMG-electrocat? Or will they have to wait for the development version of RMS that you are using to be merged into RMS main?

The idea is that they would be able to use it.  Until the next release of course, users would have to be using the "install from source" or developer docker version. 

@ssun30  please add to the checklist (and then add a commit to this PR to do it) editing the "install from source" instructions for users so that they use the appropriate `marcus_development` branch of RMS. (maybe we should pick a better name for it, @mjohnson541 ? this is essentially a branch of RMS for use by `main` RMG)

**Comment by JacksonBurns** [comment]
I am confused, the current installation instructions still say to use the version of `julia` we install from `conda-forge`, but this update requires using `julia` installed via `juliaup` (i.e. what we are doing in #2687)?

**Comment by rwest** [comment]
I think this PR doesn't change how we install Julia. (Just which branch of RMS is used). We deliberately separated out all the Julia issues to avoid PR dependency stalemate.

**Comment by rwest** [comment]
I've renamed the RMS branch that this follows to be `for_rmg` instead of `marcus_development`. We used this name previously (Summer 2023) when we temporarily needed to avoid the `main` RMS branch.  I updated the CI and documentation to match.

**Comment by ssun30** [comment]
> I've renamed the RMS branch that this follows to be `for_rmg` instead of `marcus_development`. We used this name previously (Summer 2023) when we temporarily needed to avoid the `main` RMS branch. I updated the CI and documentation to match.

The `RMG_DATABASE_BRANCH` environment variable in the CI was set at the wrong location (under the `build-osx` job) in commit [564c62c](https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2598/commits/564c62cde5c31b2d102c00946eb1730ea5bb6043) which causes the CI to fail because the `build-and-test-linux` job still uses `main`. It should be set as a global environment variable and I have amended this commit.

**Review by mjohnson541 — APPROVED** [review]

**Pull request merged**